src/stdlib_sparse_kinds.fypp

```diff
@@ -0,0 +1,894 @@
+#:include "common.fypp"
+#:set R_KINDS_TYPES = list(zip(REAL_KINDS, REAL_TYPES, REAL_SUFFIX))
+#:set C_KINDS_TYPES = list(zip(CMPLX_KINDS, CMPLX_TYPES, CMPLX_SUFFIX))
+#:set KINDS_TYPES = R_KINDS_TYPES+C_KINDS_TYPES
+!> The `stdlib_sparse_kinds` module provides derived type definitions for different sparse matrices
+!>
+! This code was modified from https://github.com/jalvesz/FSPARSE by its author: Alves Jose
+module stdlib_sparse_kinds
+    use ieee_arithmetic
+    use stdlib_kinds, only: int8, int16, int32, int64, sp, dp, xdp, qp
+    implicit none
+    private
+
+    ! -- Global parameters
+    enum, bind(C)
+        enumerator :: sparse_full  !> Full Sparse matrix (no symmetry considerations)
+        enumerator :: sparse_lower !> Symmetric Sparse matrix with triangular inferior storage
+        enumerator :: sparse_upper !> Symmetric Sparse matrix with triangular supperior storage
+    end enum
+    public :: sparse_full, sparse_lower, sparse_upper
+
+    #:for k1, t1, s1 in (R_KINDS_TYPES)
+    ${t1}$, parameter, public :: zero_${s1}$ = 0._${k1}$
+    ${t1}$, parameter, public :: one_${s1}$ = 1._${k1}$
+    #:endfor
+    #:for k1, t1, s1 in (C_KINDS_TYPES)
+    ${t1}$, parameter, public :: zero_${s1}$ = (0._${k1}$,0._${k1}$)
+    ${t1}$, parameter, public :: one_${s1}$ = (1._${k1}$,1._${k1}$)
+    #:endfor
+
+    ! -- Classes
+
+    !> Base sparse type holding the meta data related to the storage capacity of a matrix.
+    type, public, abstract :: sparse_type
+      integer :: nrows = 0 !> number of rows
+      integer :: ncols = 0 !> number of columns
+      integer :: nnz   = 0 !> number of non-zero values
+      integer :: storage = sparse_full !> assumed storage symmetry
+    end type
+
+    !> COO: COOrdinates compresed format 
+    type, public, extends(sparse_type) :: COO_type
+      logical               :: is_sorted = .false. !> whether the matrix is sorted or not
+      integer, allocatable  :: index(:,:) !> Matrix coordinates index(2,nnz)
+    contains
+      procedure :: malloc => malloc_coo
+    end type
+
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    type, public, extends(COO_type) :: COO_${s1}$
```

Comment:
Just to be consistent with rest of `stdlib`:
```suggestion
    type, public, extends(COO_type) :: COO_${s1}$_type
```

However, I will agree that it will become a bit long.